numba/targets/boxing.py

```diff
@@ -661,6 +661,179 @@ def reflect_list(typ, val, c):
         list.set_dirty(False)
 
 
+def _python_set_to_native(typ, obj, c, size, setptr, errorptr):
+    """
+    Construct a new native set from a Python set.
+    """
+    # Allocate a new native set
+    ok, inst = setobj.SetInstance.allocate_ex(c.context, c.builder, typ, size)
+    with c.builder.if_else(ok, likely=True) as (if_ok, if_not_ok):
+        with if_ok:
+            # Traverse Python set and unbox objects into native set
+            typobjptr = cgutils.alloca_once_value(c.builder,
+                                                  ir.Constant(c.pyapi.pyobj, None))
+
+            with c.pyapi.set_iterate(obj) as loop:
+                itemobj = loop.value
+                # Mandate that objects all have the same exact type
+                typobj = c.pyapi.get_type(itemobj)
+                expected_typobj = c.builder.load(typobjptr)
+
+                with c.builder.if_else(
+                    cgutils.is_null(c.builder, expected_typobj),
+                    likely=False) as (if_first, if_not_first):
+                    with if_first:
+                        # First iteration => store item type
+                        c.builder.store(typobj, typobjptr)
+                    with if_not_first:
+                        # Otherwise, check item type
+                        type_mismatch = c.builder.icmp_signed('!=', typobj,
+                                                              expected_typobj)
+                        with c.builder.if_then(type_mismatch, likely=False):
+                            c.builder.store(cgutils.true_bit, errorptr)
+                            c.pyapi.err_set_string("PyExc_TypeError",
+                                                   "can't unbox heterogenous set")
```

Comment:
That's a good question. To me `NotImplementedError` would mean we plan to support it later, but we aren't likely to ever support heterogenous collections, so I think `TypeError` is truer to the intent. But that's mostly a cultural question :-)